proxy/redirect.go

```diff
@@ -0,0 +1,96 @@
+package proxy
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/estuary/data-plane-gateway/auth"
+	"github.com/sirupsen/logrus"
+)
+
+// authRedirectHandler is used in the authentication flow for accessing private ports throught the proxy.
+// It handles requests to `/auth-redirect` endpoints after users have successfully acquired an auth token
+// from the dashboard (control-plane). It expects the request URI to contain `token` and `orig_url`
+// parameters. Assuming the request and token are valid, the user will be redirected to `orig_url`
+// with a cookie that contains their auth token.
+type authRedirectHandler struct {
+	dpgDomainSuffix string
+}
+
+func newRedirectHandler(dpgHostname string) *authRedirectHandler {
+	return &authRedirectHandler{
+		dpgDomainSuffix: "." + dpgHostname,
+	}
+}
+
+func (h *authRedirectHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	var params = req.URL.Query()
+	// Note that we don't do any validation of the token here. If it's not valid, then we'll
+	// catch it when the browser requests the new location, and can handle it then.
+	var token = params.Get("token")
+	if token == "" {
+		renderAuthError(errors.New("url is missing the token parameter"), w, req)
+		return
+	}
+	var origUrl = params.Get("orig_url")
+	if origUrl == "" {
+		renderAuthError(errors.New("url is missing the orig_url parameter"), w, req)
+		return
+	}
+	var origUrlParsed, err = url.Parse(origUrl)
+	if err != nil {
+		renderAuthError(fmt.Errorf("invalid orig_url parameter: %w", err), w, req)
+		return
+	}
+
+	// Check that the hostname of the original url is actually a subdomain of DPG's hostname.
+	// This isn't technically neccessary for security because we use a
+	if !strings.HasSuffix(origUrlParsed.Hostname(), h.dpgDomainSuffix) {
+		renderAuthError(fmt.Errorf("invalid orig_url parameter: hostname '%s' is not a subdomain of %s", origUrlParsed.Hostname(), h.dpgDomainSuffix), w, req)
```

Comment:
We talked about this on the call, but:

This will prevent the scenario of a malicious 3rd party site redirecting a user to `https://.../redirect?orig_url=malicious_domain.com` and stealing the token after a successful redirect. 

It doesn't solve the issue of someone running a malicious workload on the Flow platform (somehow, currently we don't let you run non-blessed connector images) and tricking the user into clicking a redirect link back to _their_ subdomain. But that's not really a big issue yet, and can be solved by scoping the tokens.

Reply:
Yeah, that's right. While looking at this, I noticed the cut off comment and finished that. Also, I think we can actually implement a proper validation in the UI, just by doing a lookup of all the task endpoints for the given `prefix` (`catalog_name` would have been a better name for that, in retrospect), and asserting that the hostname of `orig_url` is included in the set of endpoints. I think that would actually make this all watertight, at least to my knowledge. I'll add an issue to the UI repo.